src/FwdState.cc

```diff
@@ -964,6 +938,47 @@ FwdState::connectStart()
     AsyncJob::Start(cs);
 }
 
+/// send request on an existing connection dedicated to the requesting client
+void
+FwdState::usePinned()
+{
+    // we only handle pinned destinations; others are handled by connectStart()
+    assert(!serverDestinations.empty());
+    assert(!serverDestinations[0]);
+
+    const auto connManager = request->pinnedConnection();
+    debugs(17, 7, "connection manager: " << connManager);
+
+    // the client connection may close while we get here, nullifying connManager
+    const auto temp = connManager ? connManager->borrowPinnedConnection(request) : nullptr;
+    debugs(17, 5, "connection: " << temp);
+
+    // the previously pinned idle peer connection may get closed (by the peer)
+    if (!Comm::IsConnOpen(temp)) {
+        serverConn = nullptr;
+        const auto anErr = new ErrorState(ERR_ZERO_SIZE_OBJECT, Http::scServiceUnavailable, request);
+        fail(anErr);
+        stopAndDestroy("pinned connection failure");
```

Comment:
Why are we not handling this failure as other connection failures? Should not we call retryOrBail() instead of stopAndDestroy() here? I understand that the end result will be the same in the current PR code because checkRetry() will return false for pinned destinations. However, if somebody adds support for re-pinning to future Squid versions, checkRetry() may return true and retryOrBail() will retry (it will need to be adjusted to remove the failed pinned connection, but that adjustment would be the responsibility of whoever adds that support). It feels like the decision whether to stop or retry does not belong to _this_ method.

Reply:
This is true. 
However just calling the retryOrBail() here will not work, because the request->flags.pinned flag is not set yet. We can always set this flag earlier, before check if the pinned connection is open, I believe will work. 

Does it sounds ok to mark request->flags.pinned flag even if the pinned connection is closed? It make sense if repining is supported but not sure if make sense now.

Reply:
> IRC, request->flags.pinned means that the request is being forwarded to a PINNED (i.e. nil path) destination. If that is true, then we should set this flag at the beginning of the usePinned() method. The current pinned connection state is irrelevant -- forwarding may succeed or fail for various reasons.

OK.
> Why do you say if? Re-pinning is supported (in some cases) AFAICT -- supportsRepinning() returns true for certain transactions.

Yes you are right, sorry confused. We are allowing repining on non pconn-race cases.
And here, if I am not wrong we have not a  pconn race case (because we detected closed connection before try to send anything).

So we can try the next available path, if any?

Reply:
To re-answer your question:
> Why are we not handling this failure as other connection failures? Should not we call retryOrBail() instead of stopAndDestroy() here

Yes we should call retryOrBail here. However we need to decide if finding  a pinned connection closed is a pconn race case or not:
  -  If yes we should retry (re-pin) the same path. Because it is not supported on pinned connections this is will result to return an error. This is the current behaviour but the code will be more consistent.
  - If not we should retry on the next available path it any.

Reply:
> we need to decide if finding a pinned connection closed is a pconn race

It is not a pconn race if we were about to send the first request on that pinned connection. For example, for all practical purposes, I cannot tell the difference between

1. A peer closing a pinned connection before we send our first HTTP request on that connection
2. A peer closing a regular connection before we send our first HTTP request on that connection.


It is a pconn race if we are trying to forward another (i.e. N+1st) request over that connection. For example, for all practical purposes, I cannot tell the difference between 

1. A peer closing a pinned connection before seeing our next HTTP request on that connection
2. A peer closing a regular idle persistent connection before seeing our next HTTP request on that connection.

> If yes we should retry (re-pin) the same path

### When there is a pconn race

I believe Squid should attempt to retry the same path because that is how we handle pconn failures and there seems to be no compelling reason to make this case special. 

* If re-pinning is prohibited, then this "attempt to retry the same path" must fail and must end the whole forwarding task. 
* Otherwise, this "attempt to retry the same path" may succeed. The attempt failure in this case ought to result in Squid trying another destination (if any). The connection should be re-pinned to the destination of a successfully established connection (if any).

### When there is no pconn race

I believe Squid should attempt to try a different path because that is how we handle non-pconn failures and there seems to be no compelling reason to make this case special. 

* If re-pinning is prohibited, then this "attempt to try a different path" must fail and must end the whole forwarding task.
* Otherwise, this "attempt to try a different path" may succeed. If it does, the connection should be re-pinned accordingly.

In a sense, "re-pin" does not mean just "pin the same destination again". Instead, it means "pin again" or "pin a new connection [after a failure of a previously pinned connection]". Whether the newly pinned destination address changes is irrelevant. The current supportsRepinning() description in this PR matches this model. However, if I am wrong, and the change in destination address is actually important (e.g., true supportsRepinning() implies that the destination must or must not change) , then we must fix supportsRepinning() description (and possibly implementation and/or caller).

Please let me know if my logic is flawed or the above answers do not allow you to move this PR forward. The ball is on your side AFAICT.

Reply:
> to become pinned, the (1) connection should have had some data already transmitted which made us set it as such in the first place

I do not think the already sent low-level bytes affect what I have said -- it is "not a pconn race" for the first high-level (usually HTTP) request, whether we have already sent some lower-level (usually TLS) bytes or not.

If Squid would have to (re)send some low-level bytes on a re-pinned connection but cannot do that (for any reason), then the new supportsRepinning() method must naturally return false (and, hence, block the new/replacement connection from being opened in the first place). However, the classification I was talking about happens _before_ a supportsRepinning() call. The result of that call affects what will happen later, after that classification, probably on different code paths.

Reply:
@rousskov 
I fully agree with you pconn race cases.
Also Yes I agree with this:
 >  Instead, it means "pin again" or "pin a new connection [after a failure of a previously pinned connection]". Whether the newly pinned destination address changes is irrelevant.

I have a concern about the following:
> Why are we not handling this failure as other connection failures? Should not we call retryOrBail() instead of stopAndDestroy() here? 

I know that initially I said that yes we should use retryOrBail instead of stopAndDestroy. However trying to implement it I realised the following problem: replacing with retryOrBail is make a risk in the future for a call sequence like the following, because of a bug: 
useDestinations->usePinned(conection is closed)->retryOrBail->useDestinations->usePinned(connection is closed)->...

Currently this is not possible (I hope!) because the supportsRepinning will fail and retryOrBail will result to return an error. However on the future a small buggy change (eg on marking pconn races) may cause such bug. I am feeling safer to just call stopAndDestroy for now.

Opinion on this?

Reply:
> useDestinations -> usePinned (connection is closed) -> retryOrBail -> useDestinations -> usePinned (connection is closed) -> ...

Good point! A raceHappened retry (if allowed by supportsRepinning()) of a PINNED (i.e. nil) destination must result in opening of a new connection. Where should this new connection go? It should go to the destination address of the now-closed pinned connection. Once opened, it should be pinned back to the client, of course (I assume we do that re-pinning already somewhere).

What is the best way to accomplish the above connection opening sequence? I am not sure, but I suspect that we should have the following set of methods instead of connectStart() and usePinned():

1. usePinnedConnection() -- same as usePinned() today. Reuses the pinned connection.
1. useFirstDestinationAddress() -- same as the beginning of connectStart() today. Then calls useDestination(serverDestinations[0]).
1. usePinnedDestinationAddress() -- computes pinnedDestination (i.e. the address of the closed/failed pinned connection) and then calls useDestination(pinnedDestination).
1. useDestination(destination) -- same as the bulk of connectStart() today but uses the supplied destination address instead of serverDestinations[0].

When serverDestinations is not empty, FwdState::useDestinations() must call one of the first three methods. For example, it must call usePinnedDestinationAddress() if pconnRace is raceHappened and serverDestinations[0] is nil.

Would the above plan address your concern?

Reply:
> Good point! A raceHappened retry (if allowed by supportsRepinning()) of a PINNED (i.e. nil) destination must result in opening of a new connection. Where should this new connection go? It should go to the destination address of the now-closed pinned connection. Once opened, it should be pinned back to the client, of course (I assume we do that re-pinning already somewhere).

No, no no no no. The very concept of automatically re-opening a PINNED connection from the middleware is invalid in todays networks.

Thanks to NAT, NTLM, Negotiate and TCP load balancers existing the new TCP connection might end up going to a completely different origin server and breaking application level things in far worse ways than we can recover from. If the connection is closed or has low-level TCP errors we are left with only the option to destruct that entire transaction and both connections which were pinned. That is fundamentally what it means to be PINNED vs anything else.

There are ways to repeat or re-negotiate at higher levels *sometimes* such as re-sending a HTTP request - but some things even then can effectively prohibit retry.

Anyhow, if the point of classification/decision whether to retry cannot determine the PINNED vs non-PINNED reliably that is a bug which needs sorting out before any of the logic depending on it being done right gets touched or discussed.

We have HttpRequest (flags.pinned) and server destination (peerType == PINNED). One of which should always be available to test - if not there is a core problem.

The other side of the test is what type of error happened (TCP? TLS? HTTP?):
 - TCP errors are completely non-recoverable for PINNED. So => destruct.
 - TLS and other middle layer errors *might* be, but I don't think we have the state stored to indicate that yet. So for now => destruct.
 - HTTP requests maybe re-send - but only over the *same* TCP connection. If that was already gone/destroyed => destruct.


(Ideally we want Squid to exactly mirror the same error on the client connection and see how the client reacts as a guide to recovery. But that is a long-term goal.)

Reply:
> Thanks to NAT, NTLM, Negotiate and TCP load balancers existing the new TCP connection might end up going to a completely different origin server and breaking application level things in far worse ways than we can recover from

But why? On repin the new server will request new authentication process from client. Which is the problems we may have?

I must note that the current squid releases allow reconnect/repin pinned connections on the next available path for non pconn race cases. 

Do you believe that we should just abort pinned connections (both client-side and server side connections) on any error?

Reply:
@yadij, if you think that all re-pinning must be prohibited now, starting from this PR, then please say so explicitly. This PR introduces supportsRepinning(), a dedicated method that decides whether re-pinning is allowed. Saying "no re-pinning should be allowed ever" in a comment to _that_ method would be ideal, but you can say it in this thread instead if you must.

This PR preserves the current status quo where that status quo does not interfere with the bug fix, allowing re-pinning in many cases that were allowed before this PR. Allowing those cases is (a lot) more work for Christos, so if you are _sure_ that we should stop allowing all of them, and that we should stop allowing all of them in _this_ PR, then we can use the saved effort as an excuse to extend this PR scope from a bug fix into a policy change. Otherwise (i.e., if you are not sure about both of those things), then we can postpone this particular discussion until somebody submits a PR that prohibits more (or all) re-pinning use cases.

Reply:
@yadij, please state whether you think that all re-pinning must be prohibited starting from this PR. The next step on this PR depends on that decision. If you cannot answer, please say so as well so that @chtsanti does not wait for your answer.

My previous [comment](https://github.com/squid-cache/squid/pull/351#discussion_r250468402) on this thread has more details.

Reply:
Yes I think things using the PINNED mechanisms should never be re-pinned.

That is not to say the things using it right now fit that requirement. There may well be cases where that strict two-way death of pinning is not needed. For those I think we should use something other than the pinning mechanism we have now. Perhapse a dedicated pconn pool or such for that client.

Reply:
@chtsanti , FWIW, I am interpreting Amos' answer as "keep re-pinning as is (to the extent possible) unless you want to change the scope of this PR from a small bug fix into a rewrite of what connections are using the existing pinning mechanism and into the creation of some new mechanism for those connections that should not be using the existing pinning mechanism but are using it today."

This is just my interpretation, of course, but even if it is incorrect, the creation of a new mechanism to replace pinning for some connections can clearly be declared out of this PR scope. Moreover, the work you are doing now will not hurt whatever that future mechanism is going to be, and whatever connections will use that future mechanism. Maybe the code will be simplified further, maybe not. The person responsible for that future rewrite will find the best way forward.

Based on the above interpretation, I suggest that you proceed as before -- nothing has changed (except the time wasted).

Reply:
It does mean the words "re-pin" and "re-pinning" are invalid. The documentation and methods using them (with or without the hyphen) need to be re-worded / re-named to avoid confusion.

Reply:
> It does mean the words "re-pin" and "re-pinning" are invalid. The documentation and methods using them (with or without the hyphen) need to be re-worded / re-named to avoid confusion.

Amos has right. The resulted client-side/server-side connections pair is not pinned, we should not use the term re-pin.
We should rename the "FwdState::supportsRepinning()" to something else for example, reconnectAfterPinned() or something like that.

Reply:
@rousskov 
> Would implementing the above plan address your concern?

Still I have concerns to implement it. I believe it is safer to not retry a pinned connection when it is found closed while trying to use it  inside FwdState::usePinned().

Moreover we do not have a safe mechanism to retry a raceHappened on pinned connections case.  The new usePinnedDestinationAddress() is not safe to be used and must not exist at all to avoid confusion.
In the case of raceHappened (zeroReply and server closes the connection) on pinned connection the commClose handlers will be scheduled and will close the client-side connection too. We should add fixes to this mechanism too else it does not make sense to try reconnect on server-side.

The FwdState::usePinned() is a special case: 
 - The  connManager->borrowPinnedConnection(request) return a closed, or closing connection (server-side pinned connection is closed).
- The ConnStateData::clientPinnedConnectionRead (which monitors pinned conenction) maybe is executed and it is closed the client connection. 
- The comm close handler ConnStateData::clientPinnedConnectionClosed: (a) maybe it is already executed and will not try to close the client-side conenction , or (b) will be executed soon and in this case will be close the client-side connection because inside FwdState::usePinned we are setting the pinning.zeroReply calling the fail(anErr)  (there is a small inconsistency here). 

We must not retry this connection because again the client-side connection is closed or will be closed soon because of pinned connection handlers. I believe we must not replace the "stopAndDestroy" call with a retryOrBail call. We can add a comment here which explains the situation.

On my side I have concerns to implement a little complex mechanism just to retry failed pinned connections. If it is easy, why not, but complex thinks may result to bugs.
Probably it would be better to avoid reconnect pinned connections on  all of the retryOrBail cases and allow only reforward (forward to next path on error responses). This is what actually does the current squid.

Reply:
For now, I just add a comment to explain why the stopAndDestroy should called when found a closed pinned connection inside FwdState::usePinned(), at commit 30b957c5.

Reply:
Looking at just the supportsRepinning() and how it is used. It seems to me that it could be named pinnedCanRetry().

In regards to your question in there and earlier this thread about HTTP auth cases. IIRC Browsers acted like they cache the non-success results for NTLM at some global level (not sure about Kerberos). So issuing a 401/407 to update the clients knowledge of the type-2 token for the new server connection [needed to get a valid type-3 token] can result in the credentials stop working entirely, user popups that reject valid logins, and much PITA.

Reply:
> Looking at just the supportsRepinning() and how it is used. It seems to me that it could be named pinnedCanRetry().

This is looks OK. 

> In regards to your question in there and earlier this thread about HTTP auth cases. IIRC Browsers acted like they cache the non-success results for NTLM at some global level (not sure about Kerberos)

I see, looks good reason to avoid re-try at least pconn-races. 
Should I uses this or similar comment inside FwdState::supportsRepinning() to explain why pinned pconn-races are not retried?

Currently in squid we are retrying pinned connections falling back to the next available path only in non pconrace cases (eg when server respond with an error page).
As is the patch now this policy is not changed.

Reply:
> Currently in squid we are retrying pinned connections only in non pconrace cases.
> As is the patch now this policy is not changed.

While the current alleged Squid behavior (of repining only non-racy errors) does not make much sense to me, I recommend keeping it as is during this PR. If there are reasons to change that behavior, a dedicated PR can be used to implement those changes. Using a dedicated PR for these changes is probably a good idea because they are outside the scope of the bug you are fixing in this PR, are far from obvious, are rather risky, and may even be controversial.

Reply:
```
        // The comm close handlers of pinned connection are already scheduled
        // and if not already executed, will close the client side connection
        // soon to propagate the zero reply to the client-side.
        // Stop trying to connect here.
```

I suggest replacing the above with the following simpler text. It relies on its code context to limit the discussion to the specific use case the code is processing -- an idle pinned connection got closed. It avoids an implication that something is necessarily propagating the ERR_ZERO_SIZE_OBJECT error created above to the client. It also avoids the deprecated "client side" terminology. 

```
        // Connection managers monitor their idle pinned to-server
        // connections and close from-client connections upon seeing
        // a to-server connection closure. Retrying here is futile.
```

Either version of this comment addresses my concern that started this change request so I am marking it as resolved. If you like my version of the comment, please commit it.

Reply:
I changed the comment inside FwdState::usePinned as suggested by Alex at commit 7efd9b35
I renamed the FwdState::::supportsRepinning to FwdState::pinnedCanRetry at commit ce7152a3.

@rousskov 
> While the current alleged Squid behavior (of repining only non-racy errors) does not make much sense to me, I recommend keeping it as is during this PR.

Not sure but imagine the case:
 - server has two available paths, path1 and path2
 - client connects to squid and establish pinned connection using path1 on server
 - After some requests served on pinned connection the path1 server become non responsible and the request timeout after long time

If we do not retry to the next available path path2 then the most possible when the client reconnects to squid will try to use again path1, fail again after some timeout and then fall back to the next available path2.

In any case needs more study and better to not try find it inside this PR which solves a different bug.